Update protobuf to 0.7.1 / repackage for NPM

crypto-coder opened the pull request:
Updating the protobuf specs to use the new ones from RChain 0.7.1
Commit hash [27d722e] https://github.com/rchain/rchain/tree/dev/models/src/main/protobuf

Also, repackaging RNode.js for the first NPM release

Should resolve the following issues:
#5 - Repackage for NPM
#6 - update protos; remove Channel and Quote
#14 - Update api for required cost accounting parameters

dckc commented:
Darn; it's not tracking comments like I thought it would. Oh well.

dckc commented:
@JoshOrndorff darn... we lost the travis config when we moved from your account.

dckc commented on protobuf/RhoTypes.d.ts:
```diff
@@ -0,0 +1,1071 @@
+import * as $protobuf from "protobufjs";
```
I'm not sure we want to maintain these typescript bindings. I'm pretty sure they're not correct.

The Makefile generates it as an intermediate file.

crypto-coder replied:
Removed this file

dckc commented on RHOCore.js:
```diff
@@ -93,6 +93,9 @@ function toJSData(par /*: IPar */) /*: Json */{
       if (typeof ex.g_string !== 'undefined') {
         return ex.g_string;
       }
+      if (typeof ex.g_byte_array !== 'undefined') {
```
RHOCore doesn't have byte arrays either. **critical**

dckc replied:
I see the test code that motivates this.

I started working on integration tests that would ask rnode how `toByteArray` works and discovered you were already doing that for the hashing functions. I factored out `runAndListen` so I could import it into the tests I'm writing.

Along the way, I refactored the code not to need byte array support in `toJSData`.

https://github.com/dckc/rchain-api-1/blob/live-tests/test/testRNode.js 6a177ef

crypto-coder replied:
Removed byte arrays from toJSData.  To add byte arrays to RHOCore, is it just a matter of putting a handler in toJSData and fromJSData?

dckc replied:
We'd have to add those handlers in such a way that maintains `deepEqual(x, toJSData(fromJSData(x)))` for all JSON data x. I don't think that's possible.

dckc commented on rnodeAPI.js:
```diff
@@ -179,7 +313,12 @@ function RNode(grpc /*: typeof grpcT */, endPoint /*: { host: string, port: numb
     getIdFromUnforgeableName,
+    sha256Hash,
+    keccak256Hash,
+    blake2b256Hash,
```
These hash functions don't depend on the node. They should just be top-level functions.

**critical**

dckc replied:
well, maybe not critical... they work this way... but they don't depend on access to the network the way this API implies. Fixing this will be a breaking API change.

crypto-coder commented:
> Let's not use the same name as rholang for different functionality. `sha256HashData` or something?
> 
> And let's export the rholang-compatible functions too.

Added Rholang-compatible functions and renamed these convenience variants with a 'simplified' prefix.  Also moved these functions out of the RNode interface so they are top-level now.  New function names are:

- sha256Hash - Rholang-style, bytes-in, bytes-out
- keccak256Hash  - Rholang-style, bytes-in, bytes-out
- blake2b256Hash - Rholang-style, bytes-in, bytes-out
- simplifiedSHA256Hash - Convenience, any RHOCore-compatible data in, HEX string out
- simplifiedKeccak256Hash - Convenience, any RHOCore-compatible data in, HEX string out
- simplifiedBlake2b256Hash - Convenience, any RHOCore-compatible data in, HEX string out

Updated unit tests with your work on your fork, and added the normal vs simplified tests.

JoshOrndorff commented:
I can't figure out how to enable travis here.Admittedly, I struggle the first time as well. But I do see how to enable it on my other repos and rchain-community repos. So I'm pretty sure this is a permissions issue. I'm contributor here, but not admin.

dckc commented on package.json:
```diff
@@ -4,8 +4,10 @@
+    "check": "npm run test && npm run lint && npm run flow-check",
     "test": "node test/testRHOCore.js; node test/testRNode.js; node test/testSigning.js",
-    "integrationTest": "node rnodeAPI.js",
+    "integrationTest": "node test/testRNode.js --net",
```
both tests should be run, no?

dckc commented on test/testRNode.js:
```diff
@@ -1,139 +1,144 @@
-const grpc = require('grpc');
-
-const { RNode, RHOCore, b2h } = require('../rnodeAPI');
+/* global require, module, exports */
```
Did you use git to merge my changes in? The **blame** view has all this code attributed to you. Maybe I can fix that later, but I'd rather avoid that sort of thing.

crypto-coder replied:
Previously copy-pasted this specific change.  Now I have merged in your changes on "live-tests", so this should reflect more accurately.

dckc commented on test/testRNode.js:
```diff
@@ -1,139 +1,144 @@
+      });
+  }
+
+  function normalHashTest(test, fn, fname) {
```
There's a lot of repeated code between `simplifiedHashTest` and `normalHashTest`. Not critical in test code, but not good for maintenance.

crypto-coder replied:
Refactored / consolidated these functions.  Differentiating the tests by adding an "isNormalTest" argument to the hashTest function.

dckc commented on test/testRNode.js:
```diff
@@ -1,139 +1,144 @@
+  function normalHashTest(test, fn, fname) {
+    const returnChannel = rng().toString(36).substring(7);
+    const txt = 'test';
+    const hashProc = `@"${fname}"!("${txt}".toByteArray(), "${returnChannel}")`;
```
when we're doing the normal hash, I wouldn't expect to see `.toByteArray()`. I would think the caller knows what bytes they want to hash, so maybe `"deadbeef".hexToBytes()` rather that `"text".toByteArray()`. That way we can use something like [Test vectors for ... SHA-3](https://www.di-mgt.com.au/sha_testvectors.html) to check our work.

I suppose that's more relevant to unit tests. Do we have any unit tests for hashing? I don't see any. I guess they can come later.

crypto-coder replied:
Refactored normal tests to use 'deadbeef' hash and simplified tests to use 'testtest' string.  Also changed function calls for normal test to use hexToBytes and h2b.

dckc commented:
Please do git merge my changes in #28 thru a10c826 . Especially critical is:

  - **RHOCore: chan.quote is no more** 3a347d6

along with test data 2e92460 and test harness 464009f

Without this fix, (a) signatures aren't compatible, and (b) rnode throws exceptions when we send it mal-formed Pars

crypto-coder commented:
> Please do git merge my changes in #28 thru [a10c826](https://github.com/rchain/rchain-api/commit/a10c82692f2840f660457d2bbdaaceb648713c4d) . Especially critical is:
> 
>     * **RHOCore: chan.quote is no more** [3a347d6](https://github.com/rchain/rchain-api/commit/3a347d60f82786470770dc842d64e6623c192868)
> 
> 
> along with test data [2e92460](https://github.com/rchain/rchain-api/commit/2e92460b6da984682be5ef5b11cb2807c4ad3940) and test harness [464009f](https://github.com/rchain/rchain-api/commit/464009f0013704c07756673098c5aad57d0ba952)
> 
> Without this fix, (a) signatures aren't compatible, and (b) rnode throws exceptions when we send it mal-formed Pars

Merged in all changes from your branch "live-tests", and ran all checks / tests.  This should include all necessary differences at this point
